### PR TITLE
Implemented PyCudaOpenCLArrayContext

### PR DESCRIPTION
**Pull request by mitkotak**

### CODE REVIEW

File: arraycontext/impl/pycuda/__init__.py [hunk 1]
```diff
@@ -0,0 +1,202 @@
...
+
+        A PyOpenCL memory allocator. Can also be `None` (default) or `False` to
+        use the default allocator.
+
+    .. automethod:: __init__
+    """
+
+    def __init__(self, allocator=None):
+        import pycuda
+        import pycuda.gpuarray as gpuarray
```
Comment (kaushikcfd) [inline arraycontext/impl/pycuda/__init__.py]: No need of importing these.

File: arraycontext/impl/pycuda/__init__.py [hunk 2]
```diff
@@ -0,0 +1,202 @@
...
+
+        super().__init__()
+        self.allocator = allocator if allocator else None
```
Comment (kaushikcfd) [inline arraycontext/impl/pycuda/__init__.py]: Unnecessary 'if-else'

File: arraycontext/impl/pycuda/__init__.py [hunk 3]
```diff
@@ -0,0 +1,202 @@
...
+
+
+    def _get_fake_numpy_namespace(self):
+        from arraycontext.impl.pycuda.fake_numpy import PyCudaFakeNumpyNamespace
+        return PyOpenCudaFakeNumpyNamespace(self)
```
Comment (kaushikcfd) [inline arraycontext/impl/pycuda/__init__.py]: Typo?

File: arraycontext/impl/pycuda/__init__.py [hunk 4]
```diff
@@ -0,0 +1,202 @@
...
+
+    # {{{ ArrayContext interface
+
+    def empty(self, shape, dtype):
+        return gpuarray.empty(shape=shape, dtype=dtype,
```
Comment (kaushikcfd) [inline arraycontext/impl/pycuda/__init__.py:77]: gpuarray is not imported.

File: arraycontext/impl/pycuda/__init__.py [hunk 5]
```diff
@@ -0,0 +1,202 @@
...
+
+    def zeros(self, shape, dtype):
+        return gpuarray.zeros(shape=shape, dtype=dtype,
+                allocator=self.allocator)
+
+    def from_numpy(self, array: np.ndarray):
+        return gpuarray.to_gpu(array, allocator=self.allocator)
+
+    def to_numpy(self, array):
+        return array.get(queue=self.queue)
```
Comment (kaushikcfd) [inline arraycontext/impl/pycuda/__init__.py]:
```suggestion
        return array.get()
```

File: arraycontext/impl/pycuda/__init__.py [hunk 6]
```diff
@@ -0,0 +1,202 @@
...
+            raise RuntimeError(
+                "Unable to reason what outer_iname and inner_iname "
+                f"needs to be; all_inames is given as: {all_inames}"
+            )
+
+        if inner_iname is not None:
+            t_unit = lp.split_iname(t_unit, inner_iname, 16, inner_tag="l.0")
+        t_unit = lp.tag_inames(t_unit, {outer_iname: "g.0"})
+
+        return t_unit
```
Comment (kaushikcfd) [inline arraycontext/impl/pycuda/__init__.py]:
```suggestion
```

File: arraycontext/impl/pycuda/__init__.py [hunk 7]
```diff
@@ -0,0 +1,202 @@
...
+
+            wait_event_queue.append(evt)
+            if len(wait_event_queue) > self._wait_event_queue_length:
+                wait_event_queue.pop(0).wait()
+
+        return result
+
+    def freeze(self, array):
+        array.finish()
+        return array.with_queue(None)
```
Comment (kaushikcfd) [inline arraycontext/impl/pycuda/__init__.py]:
```suggestion
        return array
```

File: arraycontext/impl/pycuda/__init__.py [hunk 8]
```diff
@@ -0,0 +1,202 @@
...
+
+    def thaw(self, array):
+        return array.with_queue(self.queue)
```
Comment (kaushikcfd) [inline arraycontext/impl/pycuda/__init__.py]:
```suggestion
        return array
```

File: arraycontext/impl/pycuda/__init__.py [hunk 9]
```diff
@@ -0,0 +1,202 @@
...
+        return array
+
+    def tag_axis(self, iaxis, tags: Union[Sequence[Tag], Tag], array):
+        # Sorry, not capable.
+        return array
+
+    def clone(self):
+        return type(self)(self.queue, self.allocator,
+                wait_event_queue_length=self._wait_event_queue_length,
+                force_device_scalars=self._force_device_scalars)
```
Comment (kaushikcfd) [inline arraycontext/impl/pycuda/__init__.py]: Return the correct copy of `self`.

File: arraycontext/impl/pycuda/__init__.py [hunk 10]
```diff
@@ -0,0 +1,110 @@
+"""
+.. currentmodule:: arraycontext
+.. autoclass:: PyCudaOpenCLArrayContext
```
Comment (kaushikcfd) [inline arraycontext/impl/pycuda/__init__.py]: Name typo?

File: arraycontext/impl/pycuda/__init__.py [hunk 11]
```diff
@@ -0,0 +1,110 @@
...
+
+import numpy as np
+
+from pytools.tag import Tag
+
+from arraycontext.context import ArrayContext
+
+
+if TYPE_CHECKING:
+    import pyopencl
```
Comment (kaushikcfd) [inline arraycontext/impl/pycuda/__init__.py]: pycuda?

File: arraycontext/impl/pycuda/__init__.py [hunk 12]
```diff
@@ -0,0 +1,110 @@
...
+# {{{ PyCUDAArrayContext
+
+class PyCUDAArrayContext(ArrayContext):
+    """
+    A :class:`ArrayContext` that uses :class:`pycuda.gpuarray.GPUArray` instances
+    for its base array class.
+
+    .. attribute:: allocator
+
+        A PyOpenCL memory allocator. Can also be `None` (default) or `False` to
```
Comment (kaushikcfd) [inline arraycontext/impl/pycuda/__init__.py]: pycuda?

File: arraycontext/impl/pycuda/__init__.py (hunk 4 above)
Comment (kaushikcfd) [reply arraycontext/impl/pycuda/__init__.py:77]: Looks like its added.

File: arraycontext/impl/pycuda/__init__.py [hunk 13]
```diff
@@ -0,0 +1,110 @@
...
+        array.finish()
+        return array
+
+    def thaw(self, array):
+        return array
+
+    # }}}
+
+    def clone(self):
+        return type(self)(self.allocator))
```
Comment (kaushikcfd) [inline arraycontext/impl/pycuda/__init__.py]:
```suggestion
        return type(self)(self.allocator)
```

File: arraycontext/impl/pycuda/__init__.py [hunk 14]
```diff
@@ -0,0 +1,110 @@
...
+
+    def to_numpy(self, array):
+        import pycuda.gpuarray as gpuarray
+        return array.get()
+
+    def call_loopy(self, t_unit, **kwargs):
+        raise NotImplementedError('Waiting for loopy to be more capable')
+
+    def freeze(self, array):
+        array.finish()
```
Comment (kaushikcfd) [inline arraycontext/impl/pycuda/__init__.py]:
```suggestion
```

File: arraycontext/impl/pycuda/fake_numpy.py [hunk 15]
```diff
@@ -0,0 +1,154 @@
+"""
+.. currentmodule:: arraycontext
+.. autoclass:: PyOpenCLArrayContext
```
Comment (kaushikcfd) [inline arraycontext/impl/pycuda/fake_numpy.py]: Name typo?

File: arraycontext/impl/pycuda/fake_numpy.py [hunk 16]
```diff
@@ -0,0 +1,154 @@
...
+"""
+__copyright__ = """
+Copyright (C) 2020-1 University of Illinois Board of Trustees
```
Comment (kaushikcfd) [inline arraycontext/impl/pycuda/fake_numpy.py]:
```suggestion
Copyright (C) 2021 University of Illinois Board of Trustees
```

File: arraycontext/impl/pycuda/fake_numpy.py [hunk 17]
```diff
@@ -0,0 +1,154 @@
...
+
+    def less(self, x, y):
+        return rec_multimap_array_container(operator.lt, x, y)
+
+    def less_equal(self, x, y):
+        return rec_multimap_array_container(operator.le, x, y)
+
+    # }}}
+
+    def ones_like(self, ary):
```
Comment (kaushikcfd) [inline arraycontext/impl/pycuda/fake_numpy.py]: Just use pycuda.gpuarray.ones_like

File: arraycontext/pytest.py [hunk 18]
```diff
@@ -126,6 +143,15 @@ def __str__(self):
                     self.device.platform.name.strip()))
 
 
+class _PytestPyCUDAArrayContextFactory(
+        PytestPyCUDAArrayContextFactory):
+
```
Comment (kaushikcfd) [inline arraycontext/pytest.py:149]:
```python
@property
def actx_class(self):
    return PyCUDAArrayConext
```

File: arraycontext/pytest.py [hunk 19]
```diff
@@ -126,6 +143,15 @@ def __str__(self):
...
+    def __call__(self):
+        from arraycontext import PyCUDAArrayContext
+        allocator  = None
+        return self.actx_class(allocator)
```
Comment (kaushikcfd) [inline arraycontext/pytest.py]:
```suggestion
        return self.actx_class(None)
```

**Review by kaushikcfd — CHANGES_REQUESTED** [review]

File: arraycontext/impl/pycuda/__init__.py (hunk 2 above)
Comment (mitkotak) [reply arraycontext/impl/pycuda/__init__.py]: Yup resolved them thanks

File: arraycontext/fake_numpy.py [hunk 20]
```diff
@@ -145,30 +145,30 @@ def _get_fake_numpy_linalg_namespace(self):
...
 
         # limit which functions we try to hand off to loopy
-        if name in self._numpy_math_functions:
-            return multimapped_over_array_containers(loopy_implemented_elwise_func)
-        else:
-            raise AttributeError(name)
+        #if name in self._numpy_math_functions:
+           # return multimapped_over_array_containers(loopy_implemented_elwise_func)
+        #else:
+            #raise AttributeError(name)
```
Comment (kaushikcfd) [inline arraycontext/fake_numpy.py]: These parts shouldn't be changed. Please revert the changes here.

**Review by kaushikcfd — COMMENTED** [review]
Some comments left during personal meeting.

File: arraycontext/fake_numpy.py (hunk 20 above)
Comment (mitkotak) [reply arraycontext/fake_numpy.py]: Changes reverted

File: arraycontext/loopy.py [hunk 21]
```diff
@@ -70,5 +70,47 @@ def get_default_entrypoint(t_unit):
 
 # }}}
 
-
-# vim: foldmethod=marker
+class BaseFakeNumpyNamespace:
```
Comment (kaushikcfd) [inline arraycontext/loopy.py]: Rename this to `LoopyBasedFakeNumpyNamespace`.

File: arraycontext/fake_numpy.py [hunk 22]
```diff
@@ -145,30 +145,30 @@ def _get_fake_numpy_linalg_namespace(self):
...
+
+        if name in self._c_to_numpy_arc_functions:
+            from warnings import warn
+            warn(f"'{name}' in ArrayContext.np is deprecated. "
+                    "Use '{c_to_numpy_arc_functions[name]}' as in numpy. "
+                    "The old name will stop working in 2021.",
+                     DeprecationWarning, stacklevel=3)
+
+        #normalize to C names anyway
+        c_name = self._numpy_to_c_arc_functions.get(name, name)
```
Comment (kaushikcfd) [inline arraycontext/fake_numpy.py]: This should be set to fire.

File: arraycontext/loopy.py (hunk 21 above)
Comment (mitkotak) [reply arraycontext/loopy.py]: Renamed

File: arraycontext/fake_numpy.py (hunk 22 above)
Comment (mitkotak) [reply arraycontext/fake_numpy.py]: Removed arraycontext.fake_numpy.__getattr__ ->  Implemented impl.pyopencl.fake_numpy.__getattr__

**Comment by mitkotak** [comment]
Migrated changes to https://gitlab.tiker.net/inducer/arraycontext/-/merge_requests/1/commits
